File: 1 - metodosBusca/Sodoku/csharp/Sudoku/Sudoku.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Sudoku
{
    class Sudoku
    {
        int[,] matriz;
        int dimensao;
        int totalChamadasNaoRecursivas;

        public int[,] Matriz { get => matriz; set => matriz = value; }
        public int Dimensao { get => dimensao; set => dimensao = value; }
        public int TotalChamadasNaoRecursivas { get => totalChamadasNaoRecursivas; set => totalChamadasNaoRecursivas = value; }

        public Sudoku()
        {
            this.TotalChamadasNaoRecursivas = 0;
        }

        public bool popularDoArquivo(string nomeDoArquivoSudoku)
        {            
            // To read the entire file at once
            if (File.Exists(nomeDoArquivoSudoku)) //o método de classe Exists() é invocado pela classe File
            {
                //le o conteúdo do arquivo linha por linha e armazena em uma lista/vetor de string
                string[] linhas = File.ReadAllLines(nomeDoArquivoSudoku);

                this.dimensao = linhas.Length;
                this.matriz = new int[dimensao,dimensao];
                this.inicializarMatriz();

                for (int i = 0; i < linhas.Length; i++)
                {
                    for (int j = 0; j < linhas.Length; j++)
                    {
                        this.Matriz[i, j] = Int32.Parse( linhas[i].Substring(j,1));
                    }
                }
                return true;
            } else
            {
                return false;
            }
        }

        private void inicializarMatriz()
        {
            for (int i = 0; i < this.Dimensao; i++)
            {
                for (int j = 0; j < this.Dimensao; j++)
                {
                    this.Matriz[i, j] = 0;
                }
            }
        }


        public void exibirSudoku(string frase)
        {
            Console.WriteLine(frase);
            for (int i = 0; i < this.Dimensao; i++)
            {
                if (i % 3 == 0 && i != 0)
                {
                    Console.WriteLine("-------------------");
                }
                for (int j = 0; j < Dimensao; j++)
                {
                    if (j % 3 == 0 && j != 0)
                    {
                        Console.Write("|");
                    }
                    Console.Write(this.Matriz[i,j] + " ");
                }
                Console.WriteLine();
            }
        }

        private Boolean numeroEstaNaLinha(int numero, int linha)
        {
            for (int i = 0; i < this.Dimensao; i++)
            {
                if (this.Matriz[linha,i] == numero)
                {
                    return true;
                }
            }
            return false;
        }

        private Boolean numeroEstaNaColuna(int numero, int coluna)
        {
            for (int i = 0; i < this.Dimensao; i++)
            {
                if (this.Matriz[i,coluna] == numero)
                {
                    return true;
                }
            }
            return false;
        }

        private Boolean numeroEstaNoBox(int numero, int linha, int coluna)
        {
            int linhaBoxLocal = linha - linha % 3;
            int colunaBoxLocal = coluna - coluna % 3;

            for (int i = linhaBoxLocal; i < linhaBoxLocal + 3; i++)
            {
                for (int j = colunaBoxLocal; j < colunaBoxLocal + 3; j++)
                {
                    if (this.Matriz[i,j] == numero)
                    {
                        return true;
                    }
                }
            }
            return false;
        }

        private Boolean numeroEstaNoLugarCerto(int numero, int linha, int coluna)
        {
            return !numeroEstaNaLinha(numero, linha) &&
                   !numeroEstaNaColuna(numero, coluna) &&
                   !numeroEstaNoBox(numero, linha, coluna);
        }

        public Boolean resolveSudoku(int qtdChamadas)
        {
            for (int linha = 0; linha < this.Dimensao; linha++)
            {
                for (int coluna = 0; coluna < this.Dimensao; coluna++)
                {
                    if (this.Matriz[linha,coluna] == 0)
                    {
                        for (int tentandoNumero = 0; tentandoNumero <= this.Dimensao; tentandoNumero++)
                        {
                            if (numeroEstaNoLugarCerto(tentandoNumero, linha, coluna))
                            {
                                this.Matriz[linha, coluna] = tentandoNumero;

                                //this.exibirSudoku("Chamada " + qtdChamadas);
                                this.TotalChamadasNaoRecursivas++;
                                //entra agora a estratégia de força bruta - profundidade com uso de pilha recursiva
                                if (resolveSudoku(qtdChamadas + 1)) //chamada recursiva, que representa a pilha do profundidade
                                {
                                    return true;
                                }
                                else
                                {
                                    this.Matriz[linha, coluna] = 0;
                                }
                            }
                        }
                        return false;
                    }
                }
            }
            return true;
        }
    }
}

File: 1 - metodosBusca/Sudoku/python/sudoku.py
class Sudoku:
        
    def popular_do_arquivo(self, nome_arquivo_sudoku):
        try:
            arquivo = open(nome_arquivo_sudoku, "r")
            conteudo = arquivo.read()
            
            linhas = conteudo.split("\n")
            self.dimensao = len(linhas)
            self.matriz = []         

            for linha in linhas:
                self.matriz.append(linha)                
            
            arquivo.close()

            return True
        except Exception as e:
            print(e)
            return False

    def exibir_sudoku(self,frase):
        print(frase)
        for i in range(0, self.dimensao):
            if (i % 3 == 0 and i != 0):            
                print("---------------------")
            for j in range(0, self.dimensao):
                if (j % 3 == 0 and j != 0):                
                    print("|",end=" ")    
                print(self.matriz[i][j], end=" ")
            print()
        
    def numero_esta_na_linha(self, numero, linha):    
        for i in range(0, self.dimensao):       
            if (self.matriz[linha][i] == numero):            
                return True        
        return False

    def numero_esta_na_coluna(self, numero, coluna):    
        for i in range(0, self.dimensao):       
            if (self.matriz[i][coluna] == numero):            
                return True        
        return False
    
    def numero_esta_no_box(self, numero, linha, coluna):
        linhaBoxLocal = linha - (linha % 3)
        colunaBoxLocal = coluna - (coluna % 3)
        for i in range(linhaBoxLocal, linhaBoxLocal + 3):        
            for j in range(colunaBoxLocal, colunaBoxLocal + 3):
                if (self.matriz[i,j] == numero):                
                    return True
        return False


    def numero_esta_no_lugar_certo(self, numero, linha, coluna):
        return (not numero_esta_na_linha(numero, linha)) and (not numero_esta_na_coluna(numero, coluna)) and (not numero_esta_no_box(numero, linha, coluna))
    

    def resolver_sudoku(self):    
        for linha in range(0, self.dimensao):        
            for coluna in range(0, self.dimensao):                    
                if (self.matriz[linha][coluna] == 0):                
                    for tentandoNumero in range(1, self.dimensao+1):                    
                        if (numero_esta_no_lugar_certo(tentandoNumero, linha, coluna)):                        
                            self.matriz[linha][coluna] = tentandoNumero
                            
                            if (resolver_sudoku()):                
                                return True                            
                            else:                            
                                self.matriz[linha][coluna] = 0                            
                    return False        
        return True
    


